### ISSUE TITLE
Display Buttons contents with three little dots

### ISSUE THREAD

**Issue by DPhilippe4** [BOINC/boinc#1554]
All the Boinc Manager buttons are non dynamically resized taking into account the text length. Consequently, it is even sometimes difficult to guess the missing words.

Here is a French example but all languages are probably impacted:

![capture d ecran 2016-06-07 a 18 04 15](https://cloud.githubusercontent.com/assets/19802944/15866497/29f6c31e-2cdf-11e6-890e-b7b702b724a2.png)

![capture d ecran 2016-06-07 a 18 04 43](https://cloud.githubusercontent.com/assets/19802944/15866499/2b51b4bc-2cdf-11e6-9d90-8335b16f5e95.png)

**Comment by ChristianBeer** [comment]
A workaround is to hover over the buttons and a tooltip text will appear.

**Comment by Ageless93** [comment]
@CharlieFenton, are the buttons in Advanced view a set size, or can they grow when the text grows? 
Wouldn't it be better to add word wrap so the text continues on a second line?

**Comment by CharlieFenton** [comment]
Looking at the code, I see that CBOINCTaskCtrl::UpdateControls() has:
        layoutChanged = 1;
and:
    if (layoutChanged) {
        Fit ();
    }
so this should already be happening. However, it also calls:
    EllipseStringIfNeeded(pItem->m_strNameEllipsed);
wich intentionally limits the string length to:
    int maxWidth = TASKBUTTONWIDTH - 10;
where:
#define TASKPANEWIDTH ADJUSTFORXDPI(200)
#define TASKBUTTONWIDTH ADJUSTFORXDPI(TASKPANEWIDTH - 55)

If I recall correctly, we actually allowed it to expand to an unlimited width several years ago, but the result was too annoying so we limited it. I believe that the problem is that the text in the buttons can changes dynamically, for example as you select or deselect rows, and the number of buttons can even change dynamically. This can cause the "task controls" area (as we call it) to keep changing size as you select / deselect rows, which people found undesirable.

It also changes the width of the area displaying the rows, which was also annoying to people. We didn't try wrapping, but that would have a similar effect on the vertical size and button spacing of the task control area.

The change to limit the size and ellipse the text was made in a series of commits on 1/25/12, but unfortunately the commit message doesn't explain the reason.

The checkin_notes file just says:
    - MGR: Make the task pane in the advanced view and its buttons fixed sizes; 
        ellipse the button contents if needed; show full button text plus 
        description in button tooltips; don't set button labels if unchanged.

If you want to play with this to see the effect of not restricting the size, I think you can just temporarily add a "return;" on line 324 of clientgui/BOINCTaskCtrl.cpp at the start of CBOINCTaskCtrl::EllipseStringIfNeeded(wxString& s).

**Comment by CharlieFenton** [comment]
Seeing the effect of not restricting the size requires a couple more changes to the code than I thought. The attached test version of clientage/BOINCTaskCtrl.cpp will do that. 

In theory, one could eliminate the ellipsis (shortening of the labels with 3 dots) and also avoid the annoying jumping of the boundary between the button area as follows:
Measure the width of all possible button labels in the current language using GetTextExtent() and set the width of the buttons for the widest possible label. However, the number of buttons and their label text is different for each project and changes as you select different rows in the Projects or Tasks tab. And the list of projects represented by the rows in the Tasks tab changes as tasks come and go. Also, the list of projects represented by the rows in both tabs changes when you add or remove projects. So testing the width of all possible button labels is not trivial.

**Comment by CharlieFenton** [comment]
GITHub would not allow me to attach a .cpp file to this issue, so I added the ".txt" suffix. Just remove the ".txt" to try it when building the Manager. Not that this is only for you to test the effect of not restricting the size, not for production.

[BOINCTaskCtrl.cpp.txt](https://github.com/BOINC/boinc/files/1418023/BOINCTaskCtrl.cpp.txt)

**Comment by Ageless93** [comment]
The effect isn't just on the Projects and Tasks tabs, but also on the Statistics tab as the OP's images show. But thanks, I'll get the code file and use it for a build, then test if that does things differently. I'll let you know.

**Comment by CharlieFenton** [comment]
I think that we do make any change, wrapping the text onto multiple lines (changing the height of some buttons dynamically) is a simpler and better solution than changing the width of the buttons and task control area dynamically.

**Comment by Ageless93** [comment]
Okay, had to learn how to add the file into my local repository, but think that in the end I did that as now there's only two places left where the French line is too long and gets truncated to three dots: the first button in Tasks, Montrer les taches activ... and in Statistics Masquer la liste de proj...

I suspect that it's probably better for the French translator to go through again and just shorten those and leave the (mouse-over) description at full length.

**Comment by DPhilippe4** [comment]
I translate Boinc in French with Transifex. I can shorten the text.

Now: "Show active tasks" / "Montrer les tâches actives"
After shortening:  "Active tasks" / "Les tâches actives"

And description stays without modification. Understanding remains as good as ever.

There is an option in Transifex to add a limited number of characters for translation. It will be necessary to use it for all buttons.

**Comment by DPhilippe4** [comment]
The translations are updated.

**Comment by CharlieFenton** [comment]
Related to #2363.

**Comment by DPhilippe4** [comment]
And in version 7.9.3 (I don't know before), I also see a vertical wrong sizing in Preferences window. The bottom of the text is hidden.
![capture d ecran2](https://user-images.githubusercontent.com/19802944/37554001-968540e0-29d2-11e8-9dd6-d1b1a6a303e6.png)

_As an aside, you can see the English word "Title" on top, not a French translation. A translation is not applied_

**Comment by RichardHaselgrove** [comment]
In {English, Windows} the word 'Title' doesn't appear. This part of the report may be a platform problem, rather than a translation problem.

**Comment by CharlieFenton** [comment]
I see this in English also, on Mac versions 7.9.1, 7.9.2 and 7.9.3. (Actually, the bottom of the hyperlink text in BOINC 7.8.6 was slightly truncated, but not nearly as much.)

> And in version 7.9.3 (I don't know before), I also see a vertical wrong sizing in Preferences window. The bottom of the text is hidden.


This appears to be a bug in wxWidgets 3.1.0. Note the label "Title" above the wxStaticBox at the top, even though the label is specified as wxEmptyString on line 68 of clientgui/DlgAdvPreferencesBase.cpp. According to the wxWidgets documentation, passing an empty string meansd no label. The bad label reduces the height of the wxStaticBox, causing the text inside the box to not fit.

I will test whether this has been fixed in wxWidgets 3.1.1. Can someone running 7.9.3 built with wxWidgets 3.1.0 on Linux check and let us know whether the same problem exists there?

Cheers,
--Charlie

**Comment by CharlieFenton** [comment]
> In {English, Windows} the word 'Title' doesn't appear. This part of the report may be a platform problem, rather than a translation problem.

I believe Windows was built with an older version of wxWidgets.

**Comment by RichardHaselgrove** [comment]
Yes, 3.0.1 still - so yes, perhaps a wxWidgets problem rather than a platform problem.

**Comment by CharlieFenton** [comment]
Fixed in PR #2418. @DPhilippe4 thank you for reporting this.
